File: server/pkg/api/actions.go
package api

const (
	// exploration
	ActionExploreSystem = "explore.system"
	ActionExploreWorld  = "explore.world"

	// base management
	ActionCreateBase = "base.create"
	ActionRemoveBase = "base.remove"

	// base factories
	ActionCreateFactory        = "factory.create"
	ActionRebalanceFactory     = "factory.rebalance"
	ActionChangeUpgradeProject = "factory.upgrade"
	ActionContributeToUpgrade  = "factory.contribute"
	ActionTransferFactoryItems = "factory.transfer"

	// inventory management
	ActionTransferItems = "transferItems"

	// cities
	ActionFoundCity = "city.found"

	// misc
	ActionRunCheat = "cheat"
)

type ExploreSystemPayload struct {
	SystemID string `json:"systemId"`
}

type ExploreWorldPayload struct {
	WorldID string `json:"worldId"`
}

type CreateBasePayload struct {
	WorldID  string `json:"worldId"`
	TileID   int    `json:"tileId"`
	Operator string `json:"operator"`
	Name     string `json:"name"`
}
type RemoveBasePayload struct {
	BaseID int `json:"baseId"`
}

type CreateFactoryPayload struct {
	BaseID      int    `json:"baseId"`
	FactoryName string `json:"name"`
}

type UpgradeFactoryPayload struct {
	FactoryID int                              `json:"factoryId"`
	Equipment []FactoriesTableRowEquipmentPlan `json:"equipment"`
}

type RebalanceFactoryPayload struct {
	FactoryID int `json:"factoryId"`

	Plan [][]FactoriesTableRowProductionPlan `json:"plan"`
}

type ContributeToFactoryPayload struct {
	FactoryID int                `json:"factoryId"`
	Amounts   map[string]float64 `json:"amounts"`
}

type RemoveFactoryPayload struct {
	FactoryID int `json:"factoryId"`
}

type TransferFactoryItemsPayload struct {
	FactoryID         int                `json:"factoryId"`
	FromFactoryToBase bool               `json:"fromFactoryToBase"`
	Amounts           map[string]float64 `json:"amounts"`
}

type FoundCityPayload struct {
	WorldID string `json:"worldId"`
	TileID  int    `json:"tileId"`
	Name    string `json:"name"`
}

type RunCheatPayload struct {
	Cmd string `json:"cmd"`
}

File: ui/src/store/datafront/actions.ts
import { createDatafrontAction } from '@/lib/datafront/action';
import {
    type ExploreSystemPayload,
    ActionExploreSystem,
    type ExploreWorldPayload,
    ActionExploreWorld,
    type FoundCityPayload,
    ActionFoundCity,
    type CreateBasePayload,
    ActionCreateBase,
    type RemoveBasePayload,
    ActionRemoveBase,
    type RunCheatPayload,
    ActionRunCheat,
    type CreateFactoryPayload,
    ActionCreateFactory,
    type RebalanceFactoryPayload,
    ActionRebalanceFactory,
    type UpgradeFactoryPayload,
    ActionChangeUpgradeProject,
    type ContributeToFactoryPayload,
    ActionContributeToUpgrade,
    type TransferFactoryItemsPayload,
    ActionTransferFactoryItems,
    type RenameFactoryPayload,
    ActionRenameFactory,
    type RenameBasePayload,
    ActionRenameBase,
    type DemolishFactoryPayload,
    ActionDemolishFactory,
} from '@/lib/net/types.generated';
import { ws } from '@/lib/net/ws';

export const dfExploreSystem = createDatafrontAction<ExploreSystemPayload>({ name: ActionExploreSystem, ws });
export const dfExploreWorld = createDatafrontAction<ExploreWorldPayload>({ name: ActionExploreWorld, ws });

export const dfFoundCity = createDatafrontAction<FoundCityPayload>({ name: ActionFoundCity, ws });

export const dfCreateBase = createDatafrontAction<CreateBasePayload>({ name: ActionCreateBase, ws });
export const dfRenameBase = createDatafrontAction<RenameBasePayload>({ name: ActionRenameBase, ws });
export const dfRemoveBase = createDatafrontAction<RemoveBasePayload>({ name: ActionRemoveBase, ws });

export const dfCreateFactory = createDatafrontAction<CreateFactoryPayload>({ name: ActionCreateFactory, ws });
export const dfRebalanceFactory = createDatafrontAction<RebalanceFactoryPayload>({ name: ActionRebalanceFactory, ws });
export const dfUpgradeFactory = createDatafrontAction<UpgradeFactoryPayload>({ name: ActionChangeUpgradeProject, ws });
export const dfContributeToFactory = createDatafrontAction<ContributeToFactoryPayload>({
    name: ActionContributeToUpgrade,
    ws,
});
export const dfTransferFactoryItems = createDatafrontAction<TransferFactoryItemsPayload>({
    name: ActionTransferFactoryItems,
    ws,
});
export const dfRenameFactory = createDatafrontAction<RenameFactoryPayload>({
    name: ActionRenameFactory,
    ws,
});
export const dfDemolishFactory = createDatafrontAction<DemolishFactoryPayload>({
    name: ActionDemolishFactory,
    ws,
});

export const dfRunCheat = createDatafrontAction<RunCheatPayload, unknown>({ name: ActionRunCheat, ws });
